api/v1alpha8/openstackservergroup_types.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha8
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// ServerGroupFinalizer allows ReconcileOpenStackServerGroup to clean up OpenStack resources associated with OpenStackServerGroup before
+	// removing it from the apiserver.
+	ServerGroupFinalizer = "openstackservergroup.infrastructure.cluster.x-k8s.io"
+)
+
+// OpenStackServerGroupSpec defines the desired state of OpenStackServerGroup
+type OpenStackServerGroupSpec struct {
+	// The name of the cloud to use from the clouds secret
+	// +optional
+	CloudName string `json:"cloudName"`
```

Comment:
This seems a bit weird, we should probably have a reference to an OpenStackCluster instead?

Reply:
Thank you for the feedback! Yeah, this allows the resource to be reconciled alone, as it's self contained.

However that isn't in any of the use cases, it doesn't seem a limitation to be tied to an existing OpenStackCluster even if the OpenStackServerGroup was only used in MachineSets. It would remove duplication of these creds.

I'll make this change, once the CRD approach is agreed.

Reply:
Hm, okay, that's a fair point. The use case to keep all the workers from different clusters in a single ServerGroup makes sense, I see your point.

Reply:
In `v1beta1` CloudName has moved into IdentifyRef, but this review thread is still relevant.

For now, I've chosen not to tie the OpenStackServerGroup to an OpenStackCluster reference and instead hold it's own secret reference. This matches how OpenStackMachine's are designed, and feels reasonable for this resource and avoids circular dependency.

I don't think there's much of a use case for two clusters using a single `OpenStackServerGroup`, that's not the design intention (but is possible with the current implementation).